stdlib/public/core/Collection.swift

```diff
@@ -260,9 +260,10 @@ extension CollectionType {
 
   /// Returns the first element of `self`, or `nil` if `self` is empty.
   ///
-  /// - Complexity: O(1)
+  /// - Complexity: O(1) in most cases, possibly O(N) for lazy collections.
   public var first: Generator.Element? {
-    return isEmpty ? nil : self[startIndex]
+    var gen = generate()
+    return gen.next()
```

Comment:
An alternative implementation looks like

```swift
let start = startIndex
return start == endIndex ? nil : self[start]
```

This basically just reimplements `isEmpty` locally without throwing away the start index. The generator approach seems slightly more straightforward though (and with my proposal to add `var first` to `SequenceType`, this is how it would be implemented there anyway).